### DISCUSSION TITLE
ExtensionObject in MethodArgument empty

### DISCUSSION THREAD

**Discussion by BenediktEnger** [FreeOpcUa/opcua-asyncio#896] [Q&A] [answered]
I have a method with one input argument and one output argument. Both arguments are custom extension objects.
When trying to call the method with UAExpert, the input and outputfields are empty:

![image](https://user-images.githubusercontent.com/44339476/169497932-393e18c8-cd37-45cd-a767-8f208f61bbe1.png)

I checked the linked datatype NodeID and this one is right. 
Here you see my method generation: 
```
def __addMethod__(self, member, parentObject):
        inputArgs = []
        outputArgs = []
        self.server.server.load_data_type_definitions()

        for input in member.inputArgs:
            
            if issubclass(type(input.type), DataTypeType):
                dataType = input.type.addStruct(self.server)
            else:
                dataType = ua.NodeId(input.type.value,0)

            arg = ua.Argument()
            arg.Name = input.name
            arg.DataType = dataType
            arg.ValueRank = -1
            arg.ArrayDimensions = []
            inputArgs.append(arg)

        for output in member.outputArgs:
            
            if issubclass(type(output.type), DataTypeType):
                dataType = output.type.addStruct(self.server)
            else:
                dataType = ua.NodeId(output.type.value,0)
            
            arg = ua.Argument()
            arg.Name = output.name
            arg.DataType = dataType
            arg.ValueRank = -1
            arg.ArrayDimensions = []
            outputArgs.append(arg)

        var = parentObject.add_method(
            ua.NodeId(self.nodeId + "/" + member.name, self.server.idx),
            ua.QualifiedName(member.name, self.server.idx),
            self.GenericUAMethod,
            inputArgs,
            outputArgs
        )

        return var
```

and this here you can see how I generate the extension object:

```
 def addStruct(self, server):

        for typeClass, nodeID in ua.datatype_by_extension_object.items():  
            if typeClass.__name__ == self.typeName:
                return nodeID

        structFields = []
        for member in self.members:
            if issubclass(type(member.type), DataTypeType):
                nodeid = member.type.addStruct(server)
                structFields.append(new_struct_field(member.name, nodeid))
            else:
                structFields.append(new_struct_field(member.name, member.type))

        struct = new_struct(server.server, server.idx, self.typeName,structFields)

        server.dict_builder.set_dict_byte_string()
        server.server.load_data_type_definitions()
        return struct[0].nodeid
```

**Comment by schroeder-** [comment, marked as answer]
Can you look if your datatypes contains any values in UAExpert? To do this, go 
Types->Datatypes->BaseDataType->Structures->[YourDatatype] and post the content of DataTypeDescription:
Like this: 
![image](https://user-images.githubusercontent.com/1273483/169702571-d954b605-c36c-40b8-a83f-328c190df1b9.png)

**Reply by BenediktEnger** [reply]
There are no values shown:

![image](https://user-images.githubusercontent.com/44339476/169752296-02e5573c-8258-462e-9070-7beb1d1e94f6.png)

**Reply by schroeder-** [reply]
Then the generated struct has no fields and ua expert is correctly showing a empty struct. There must be an error in the structfield generation code.

**Reply by BenediktEnger** [reply]
For testing, I added a struct with a single field (Float). 
I debugged it to see what is happening but in my opinion, it looks good. a structField is getting created and added to an array which I pass to newStruct:

![image](https://user-images.githubusercontent.com/44339476/169769862-1a6cc17d-d087-46f7-9054-f194ea06f10b.png)

**Reply by schroeder-** [reply]
Maybe setting the logging level with DEBUG around your code, gives a clue whats wrong.

**Reply by BenediktEnger** [reply]
There are no additional info showing up when I set the logging level to DEBUG

**Reply by BenediktEnger** [reply]
I used the example for custom structures in this repository and there, it looks the same:

![image](https://user-images.githubusercontent.com/44339476/169968913-18fe3d11-06fe-4a99-add8-a7b8a7520771.png)

**Reply by BenediktEnger** [reply]
Ok it seems it is not working at all. I combined two examples (extension objects and methods with extension object arguments) and I get the same result as you can see below:

![image](https://user-images.githubusercontent.com/44339476/169971613-63d14734-126e-44db-b120-dff7ad8fb35c.png)

![image](https://user-images.githubusercontent.com/44339476/169971812-df607189-a4c7-478f-b979-e58ca4cc5568.png)

**Reply by schroeder-** [reply]
Which version of UAExpert and asyncua are you using?

**Reply by BenediktEnger** [reply]
I just updated to the latest UAExpert and now it works. Thank you for your time!
